### PR TITLE
Add tests to validate: correctly escape '#' character in key & label paths and queries

### PR DESCRIPTION
**Pull request by annelo-msft**
Implements issue #8073.

### CODE REVIEW

**Review by pakrym — APPROVED** [review]

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs [hunk 1]
```diff
@@ -874,6 +935,245 @@ public async Task GetBatchSettingWithAllFields()
...
+
+                ConfigurationSetting[] settings = (await service.GetSettingsAsync(selector, CancellationToken.None).ToEnumerableAsync()).ToArray();
+
+                Assert.GreaterOrEqual(settings.Length, 2);
+                Assert.IsTrue(settings.Any(s => s.Key == "ab,cd"));
+                Assert.IsTrue(settings.Any(s => s.Key == "wx,yz"));
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e.Message);
```
Comment (AlexanderSher) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: What's the point of catching exception in the test and swallowing its stack trace?

Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: Nice catch, thanks!  I had it there for debugging.  Will fix.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs [hunk 2]
```diff
@@ -874,6 +935,245 @@ public async Task GetBatchSettingWithAllFields()
...
+
+            ConfigurationSetting abcSetting = new ConfigurationSetting("ab,cd", "comma in key");
+            ConfigurationSetting xyzSetting = new ConfigurationSetting("wx,yz", "comma in key");
+
+            try
+            {
+                await service.SetAsync(abcSetting);
+                await service.SetAsync(xyzSetting);
+
+                var selector = new SettingSelector(key:null);
```
Comment (AlexanderSher) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: `new SettingSelector(key:null)` is equal to `new SettingSelector(SettingSelector.Any)`:
```cs
        public SettingSelector(string key, string label = default)
        {
            Keys = new List<string>
            {
                key ?? Any
            };

            Labels = new List<string>();
            if (label != null)
                Labels.Add(label);
        }
```
You probably want to have this:
```cs
var selector = new SettingSelector("ab,cd");
selector.Keys.Add("wz,yz");
```

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs [hunk 3]
```diff
@@ -874,6 +935,245 @@ public async Task GetBatchSettingWithAllFields()
...
+                selector.Keys.Add("ab,cd");
+                selector.Keys.Add("wz,yz");
+
+                ConfigurationSetting[] settings = (await service.GetSettingsAsync(selector, CancellationToken.None).ToEnumerableAsync()).ToArray();
+
+                Assert.GreaterOrEqual(settings.Length, 2);
```
Comment (AlexanderSher) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: Shouldn't it be `Assert.Equal(settings.Length, 2);`? What other settings do we expect to be returned?

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs [hunk 4]
```diff
@@ -874,6 +935,241 @@ public async Task GetBatchSettingWithAllFields()
...
+                Assert.IsTrue(settings.Any(s => s.Label == "abc"));
+                Assert.IsTrue(settings.Any(s => s.Label == "xyz"));
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e.Message);
```
Comment (AlexanderSher) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: Another `catch`

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs [hunk 5]
```diff
@@ -874,6 +935,241 @@ public async Task GetBatchSettingWithAllFields()
...
+                var selector = new SettingSelector("abc");
+                selector.Keys.Add("xyz");
+
+                ConfigurationSetting[] settings = (await service.GetSettingsAsync(selector, CancellationToken.None).ToEnumerableAsync()).ToArray();
+
+                Assert.GreaterOrEqual(settings.Length, 2);
+                Assert.IsTrue(settings.Any(s => s.Key == "abc"));
+                Assert.IsTrue(settings.Any(s => s.Key == "xyz"));
+            }
+            catch (Exception e)
```
Comment (AlexanderSher) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: And another one

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs [hunk 6]
```diff
@@ -48,6 +50,21 @@ private ConfigurationSetting CreateSetting()
...
+            {
+                Key = GenerateKeyId($"{key}-"),
+                Value = value,
+                Label = label,
+            };
+        }
+
+        private ConfigurationSetting CreateSettingSpecialCharacters()
+        {
+            return CreateSetting($"key-{specialChars}", $"value-{specialChars}", $"label-{specialChars}");
```
Comment (AlexanderSher) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: Not important, but `key-` prefix is added twice: `CreateSetting` also adds `key-`

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs (hunk 3 above)
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: I'd like to defend against any settings that might not have been deleted from earlier runs.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs [hunk 7]
```diff
@@ -874,6 +935,241 @@ public async Task GetBatchSettingWithAllFields()
...
+            try
+            {
+                await service.SetAsync(testSetting);
+
+                var selector = new SettingSelector(testSetting.Key);
+
+                ConfigurationSetting[] settings = (await service.GetSettingsAsync(selector, CancellationToken.None).ToEnumerableAsync()).ToArray();
+
+                // There should be at least one key available
+                Assert.GreaterOrEqual(settings.Length, 1);
```
Comment (AlexanderSher) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: You can use `CollectionAssert.IsNotEmpty` here.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs [hunk 8]
```diff
@@ -874,6 +935,228 @@ public async Task GetBatchSettingWithAllFields()
...
+            try
+            {
+                await service.SetAsync(abcSetting);
+                await service.SetAsync(xyzSetting);
+
+                var selector = new SettingSelector($"{abcSetting.Key},{xyzSetting.Key}");
+
+                ConfigurationSetting[] settings = (await service.GetSettingsAsync(selector, CancellationToken.None).ToEnumerableAsync()).ToArray();
+
+                Assert.GreaterOrEqual(settings.Length, 0);
```
Comment (AlexanderSher) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationLiveTests.cs]: This assert will always succeed. Did you mean `CollectionAssert.IsEmpty(settings);`?

**Review by AlexanderSher — APPROVED** [review]

**Pull request merged**